### ISSUE TITLE
Rename use_std feature to just std

### ISSUE THREAD

**Issue by dtolnay** [rust-lang/libc#657]
As recommended by https://github.com/brson/rust-api-guidelines#c-feature. This aligns better with the naming of implicit features inferred by Cargo.

**Comment by roblabla** [comment]
What's the use-case for the `use_std` flag ? Couldn't the crate always be no_std? The only use of that feature flag is at https://github.com/rust-lang/libc/blob/master/src/unix/mod.rs#L286. What happens if we remove that check?

**Comment by alexcrichton** [comment]
Changing it can be a breaking change in some subtle situations so it wasn't turned off by deafult, but for the next major release it'll be turned off by default.

**Comment by gnzlbg** [comment]
Since this feature is enabled by default, this would only be a breaking change iff `--no-default-features --features use_std` is being used, which would be a weird configuration, since doing nothing has the same effect. 

I have a proposal here. 

We could add a `std` feature, in parallel with `use_std`, that is also enabled by default. Then, in the `build.rs`, we could translate `use_std` to `std`, and start warning on its usage. Internally, in the library, we would just use `std` as a feature.

We might not be able to remove `use_std` ever (I know a lot of crates re-exporting `use_std = ["libc/use_std"]`), but we can at least already implement a proper migration path, and start telling users to migrate to it. 

The question is whether doing this is worth it or not.

**Comment by gnzlbg** [comment]
Now the `std` feature can be used as a `cargo` feature, and the `use_std` feature has been deprecated (produces a cargo warning). I'm repurposing this issue to track the removal of the `use_std` feature, which would be a breaking change that will probably have to wait till the next breaking libc release.

**Comment by mati865** [comment]
Is `std` feature necessary?
It caused massive issues upgrading `rand` in Rust repo: https://github.com/rust-lang/rust/pull/61393
More details at: https://github.com/rust-random/rand/issues/852

Short summary:
Somehow when building Rust newer `rand` versions which use more of `libc` were failing during linking of `rustc_main`:
```
/usr/bin/ld: /usr/lib/gcc/x86_64-linux-gnu/5/../../../x86_64-linux-gnu/Scrt1.o: undefined reference to symbol '__libc_start_main@@GLIBC_2.2.5'
//lib/x86_64-linux-gnu/libc.so.6: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
```
It was found this can be avoided by disabling `std` feature of `libc` causing it to pull `libc.so` and Rust built fine.

While it looks like rustbuild issue what is the point of not linking to C library by default? AFAIK it would not hurt if C library was linked twice, by libstd and libc.

**Comment by gnzlbg** [comment]
Currently, there are three cases:

* building libc when libstd is linked,
* building libc when libstd is not linked.
    * building libc for the std library

The `std` feature should link libstd (although this is not happening anymore? libc is `#![no_std]`?), and IIUC should only be disabled for `#![no_std]` binaries or when building libstd itself. 

I don't know why `std` is different from `rustc-dep-of-std`, AFAICT both should be equal. The `rustc-dep-of-std` feature requires `no_core`, but I don't know why this is the case, since `libcore` is always built when trying to build `libc` for `libstd`.

I think maybe we can unify these two. Have `std` always enabled by default, and have `libstd` disable it.

> Is std feature necessary? While it looks like rustbuild issue what is the point of not linking to C library by default? AFAIK it would not hurt if C library was linked twice, by libstd and libc.

Does this hold for statically linked C libraries as well ? How does this interact with the crt-static target feature?

IIRC one shall not link the C libraries statically twice, and that's what the `std` feature avoided.

**Comment by mati865** [comment]
`std` feature explicitly skips linking of C standard library: https://github.com/rust-lang/libc/blob/7c8e397c0cc463e60345183d77de8b66431d7316/src/unix/mod.rs#L296-L299

If std binary depends on `#[no_std]` crate using `libc = { version = "0.2", default-features = false }` shouldn't it be considered as linking C standard library multiple times?

Although it may be oversimplified this example happen to work fine with dynamically linked glibc and statically linked musl: https://gist.github.com/mati865/b422b41f2c2784361091db78354d4bc8
Don't know about other targets.

**Comment by highjeans** [comment]
This seems to already be done with commit 4364c3b as I cannot find any references of 'use_std' anywhere in the codebase.

**Comment by dybucc** [comment]
@tgross35 does this need any solving? as per the last few comments, it seems solved.

**Comment by tgross35** [comment]
Thanks, this is indeed done. We can remove the `std` feature as well with https://github.com/rust-lang/libc/issues/5265 but that's a separate issue.
